src/manifold/src/edge_op.cpp

```diff
@@ -483,12 +484,15 @@ void Manifold::Impl::CollapseEdge(const int edge) {
   RemoveIfFolded(start);
 }
 
-void Manifold::Impl::RecursiveEdgeSwap(const int edge) {
+void Manifold::Impl::RecursiveEdgeSwap(const int edge,
+                                       std::vector<uint32_t>& visited) {
   VecDH<TriRef>& triRef = meshRelation_.triRef;
 
   if (edge < 0) return;
   const int pair = halfedge_[edge].pairedHalfedge;
   if (pair < 0) return;
+  if (visited[edge] == pair) return;
+  visited[edge] = pair;
```

Comment:
Would it be safer to mark these as booleans and set both the `[edge]` and `[pair]` to visited?

Reply:
It will likely output more degenerate tris, but I’ll try.
Otherwise maybe tagging `visited[pair] = edge`.
Ultimately, slight changes in the logic gives different result for the ‘Sponge4’ test

Reply:
Yeah, I'm not super worried about the variations in Sponge4 - really I need to come back to it and figure out why it has so many in the first place - I had managed to get rid of most of them at one time. 

Curious what you find here - I need to look into this logic more deeply at some point. It's been awhile.

Reply:
Found a way to get the same number of degenerate Tris as master, while avoiding recursion.
Instead of booleans I use tag that I reset when there is a `CollapseEdge`.

---
Alternative version: (in case the current solution skips valid case)
Using a stack and hash pair of edges instead of checking both edges separately (I don't think it's necessary)

```diff
diff --git a/src/manifold/src/edge_op.cpp b/src/manifold/src/edge_op.cpp
index 26fd983..e1c2e73 100644
--- a/src/manifold/src/edge_op.cpp
+++ b/src/manifold/src/edge_op.cpp
@@ -222,12 +222,11 @@ void Manifold::Impl::SimplifyTopology() {
                      precision_};
     for_each_n(policy, countAt(0), nbEdges,
                [=] __host__ __device__(int i) { bflagsPtr[i] = se(i); });
-    std::vector<int> visited(halfedge_.size(), -1);
-    int tag = 0;
+    std::vector<glm::ivec2> visited(halfedge_.size(), glm::ivec2(-1));
     for (int i = 0; i < nbEdges; ++i) {
       if (bflags[i]) {
-        tag++;
-        RecursiveEdgeSwap(i, visited, tag);
+        visited.resize(0);
+        RecursiveEdgeSwap(i, visited);
         numFlagged++;
       }
     }
@@ -487,7 +486,7 @@ void Manifold::Impl::CollapseEdge(const int edge) {
 }
 
 void Manifold::Impl::RecursiveEdgeSwap(const int edge,
-                                       std::vector<int>& visited, int& tag) {
+                                       std::vector<glm::ivec2>& visited) {
   VecDH<TriRef>& triRef = meshRelation_.triRef;
 
   if (edge < 0) return;
@@ -495,7 +494,9 @@ void Manifold::Impl::RecursiveEdgeSwap(const int edge,
   if (pair < 0) return;
 
   // avoid infinite recursion
-  if (visited[edge] == tag && visited[pair] == tag) return;
+  for (int i = 0; i < visited.size(); ++i) {
+    if (visited[i] == glm::ivec2(edge, pair)) return;
+  }
 
   const glm::ivec3 tri0edge = TriOf(edge);
   const glm::ivec3 tri1edge = TriOf(pair);
@@ -575,15 +576,14 @@ void Manifold::Impl::RecursiveEdgeSwap(const int edge,
     SwapEdge();
     const glm::vec2 e23 = v[3] - v[2];
     if (glm::dot(e23, e23) < precision_ * precision_) {
-      tag++;
+      visited.resize(0);
       CollapseEdge(tri0edge[2]);
     } else {
-      visited[edge] = tag;
-      visited[pair] = tag;
-      RecursiveEdgeSwap(tri0edge[0], visited, tag);
-      RecursiveEdgeSwap(tri0edge[1], visited, tag);
-      RecursiveEdgeSwap(tri1edge[0], visited, tag);
-      RecursiveEdgeSwap(tri1edge[1], visited, tag);
+      visited.emplace_back(glm::ivec2(edge, pair));
+      RecursiveEdgeSwap(tri0edge[0], visited);
+      RecursiveEdgeSwap(tri0edge[1], visited);
+      RecursiveEdgeSwap(tri1edge[0], visited);
+      RecursiveEdgeSwap(tri1edge[1], visited);
     }
     return;
   } else if (CCW(v[0], v[3], v[2], precision_) <= 0 ||
@@ -592,10 +592,9 @@ void Manifold::Impl::RecursiveEdgeSwap(const int edge,
   }
   // Normal path
   SwapEdge();
-  visited[edge] = tag;
-  visited[pair] = tag;
-  RecursiveEdgeSwap(halfedge_[tri0edge[1]].pairedHalfedge, visited, tag);
-  RecursiveEdgeSwap(halfedge_[tri1edge[0]].pairedHalfedge, visited, tag);
+  visited.emplace_back(glm::ivec2(edge, pair));
+  RecursiveEdgeSwap(halfedge_[tri0edge[1]].pairedHalfedge, visited);
+  RecursiveEdgeSwap(halfedge_[tri1edge[0]].pairedHalfedge, visited);
 }
 
 void Manifold::Impl::SplitPinchedVerts() {
diff --git a/src/manifold/src/impl.h b/src/manifold/src/impl.h
index dc4f57b..cb890df 100644
--- a/src/manifold/src/impl.h
+++ b/src/manifold/src/impl.h
@@ -121,7 +121,7 @@ struct Manifold::Impl {
   void SimplifyTopology();
   void DedupeEdge(int edge);
   void CollapseEdge(int edge);
-  void RecursiveEdgeSwap(int edge, std::vector<int>& visited, int& tag);
+  void RecursiveEdgeSwap(int edge, std::vector<glm::ivec2>& visited);
   void RemoveIfFolded(int edge);
   void PairUp(int edge0, int edge1);
   void UpdateVert(int vert, int startEdge, int endEdge);
```

Reply:
Interesting, I tend to prefer to simpler version you have now, though I don't quite grok the difference. If you want, it's probably best to put this as a follow-on PR and add a test that shows the difference.